### PR TITLE
Avoid blocking built-in node modules when using --no-bundle flag

### PR DESCRIPTION
**Pull request by richardscarrott**
**What this PR solves / how to test:**

We want to import and use the [`AsyncLocalStorage` polyfil](https://developers.cloudflare.com/workers/runtime-apis/nodejs/asynclocalstorage/) in our Remix app, deployed to CF Pages.

However, to ensure Remix sourcemaps actually align with the executed code, we tell wrangler not to bundle our code and instead use [Advanced Mode](https://developers.cloudflare.com/pages/platform/functions/advanced-mode/) (_worker.js file).

```
npx wrangler pages deploy dist --project-name="${{ inputs.cloudflare_app }}" --branch="${{ inputs.cf_branch }}" --no-bundle
```

This fails with:

> _worker.js is not being bundled by Wrangler but it is importing from another file. This will throw an error if deployed. You should bundle the Worker in a pre-build step, remove the import if it is unused, or ask Wrangler to bundle it by setting`--bundle`

Screenshot 2023-10-29 at 18 09 41

<br />
<br />

After some [discussion in Discord](https://discord.com/channels/595317990191398933/789155108529111069/1168242491741450342), I believe this is a mistake because `node:async_hooks` _is_ available to import at runtime and doesn't depend on a build step.

**Author has addressed the following:**

- Tests
  - [ ] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] Included
  - [ ] Not necessary because:
- Associated docs
  - [ ] Issue(s)/PR(s):
  - [ ] Not necessary because:

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (Skye-31) [inline packages/wrangler/src/pages/functions/buildWorker.ts:365]: Can you also check for `cloudflare:` prefixes 🙂

**Review by petebacondarwin — DISMISSED** [review]
DONT MERGE - just testing approval

**Review by Skye-31 — COMMENTED** [review]
Thanks for the PR!

Could you add a test (see https://github.com/cloudflare/workers-sdk/blob/main/fixtures/pages-workerjs-app/tests/index.test.ts#L8), as well as a changeset (`pnpx changeset` in the root of the repo)

**Comment by penalosa** [comment]
Apologies for the delay here—this is looking good! Before merging, could you add a unit test to https://github.com/cloudflare/workers-sdk/tree/main/packages/wrangler/src/__tests__/pages/deploy.test.ts ? I don't think this change will need a full fixture test

**Comment by lrapoport-cf** [comment]
hi @richardscarrott :) could you please add a unit test (per @penalosa 's [comment](https://github.com/cloudflare/workers-sdk/pull/4303#issuecomment-1850417670))? else we'll close the PR and you can re-open when you're able to revisit and make the relevant changes. thanks!

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (petebacondarwin) [inline packages/wrangler/src/pages/functions/buildWorker.ts:367]: OK so I think we just need a bit of care here.
We should only consider imports beginning with `node:` as external if the `nodejs_compat` flag is actually turned on.

We can pass that through to this plugin by making it a function with a parameter, I believe.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
One small comment but otherwise LGTM!
Thanks

**Review by petebacondarwin — APPROVED** [review]
I rebased, swapped out the fixture tests for unit tests, and added the check for the nodejs_compat flag.

@penalosa - PTAL

File: packages/wrangler/src/__tests__/pages/deploy.test.ts
Comment (penalosa) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts:2961]: This comment seems out of place?

**Review by penalosa — APPROVED** [review]
One misplaced comment, but LGTM

**Pull request merged**
